File: Examples/Comparison/pairwise_affine_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function build_sequences()
    left = join(("ACGT"[(mod(index * 7 + 3, 4) + 1)] for index in 1:1200))
    right = left[1:500] * left[571:end]
    return left, right
end

function main()
    left, right = build_sequences()
    left_record = SeqRecordLite(left; identifier="left", description="affine benchmark left sequence")
    right_record = SeqRecordLite(right; identifier="right", description="affine benchmark right sequence")

    BioToolkit.pairwise_align(left_record, right_record; match=2, mismatch=-1, gap_open=-5, gap_extend=-1)

    alignment_ms, alignment = elapsed_ms() do
        BioToolkit.pairwise_align(left_record, right_record; match=2, mismatch=-1, gap_open=-5, gap_extend=-1)
    end

    python_output = read(`conda run -n general --no-capture-output python $(joinpath(@__DIR__, "pairwise_affine_compare.py"))`, String)
    python_score = parse(Float64, match(r"python_pairwise_affine_score=(-?[0-9.]+)", python_output).captures[1])
    python_matches = parse(Int, match(r"python_pairwise_affine_matches=(\d+)", python_output).captures[1])
    python_identity = parse(Float64, match(r"python_pairwise_affine_identity=([0-9.]+)", python_output).captures[1])
    python_length = parse(Int, match(r"python_pairwise_affine_length=(\d+)", python_output).captures[1])

    @assert isapprox(Float64(alignment.score), python_score; atol=1e-8)
    @assert alignment.matches == python_matches
    @assert isapprox(round(alignment.identity, digits=6), python_identity; atol=1e-12)
    @assert length(alignment.left) == python_length

    println("Affine-gap alignment benchmark")
    println("  left_length=", length(left))
    println("  right_length=", length(right))
    println("  julia_affine_ms=", round(alignment_ms, digits=4))
    println("  julia_affine_score=", alignment.score)
    println("  julia_affine_identity=", round(alignment.identity, digits=6))
    print(python_output)
end

main()

File: Examples/Comparison/pairwise_affine_compare.py
from time import perf_counter

from Bio.Align import PairwiseAligner


def build_sequences():
    left = "".join("ACGT"[(index * 7 + 3) % 4] for index in range(1, 1201))
    right = left[:500] + left[570:]
    return left, right


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def main():
    left, right = build_sequences()

    aligner = PairwiseAligner()
    aligner.mode = "global"
    aligner.match_score = 2
    aligner.mismatch_score = -1
    aligner.open_gap_score = -5
    aligner.extend_gap_score = -1

    list(aligner.align(left, right))
    py_ms, alignments = time_call(lambda: list(aligner.align(left, right)))
    best = alignments[0]
    aligned_left = str(best[0])
    aligned_right = str(best[1])
    score = best.score
    matches = sum(1 for left_byte, right_byte in zip(aligned_left, aligned_right) if left_byte == right_byte)
    identity = matches / len(aligned_left) if aligned_left else 0.0

    print(f"python_pairwise_affine_ms={py_ms:.4f}")
    print(f"python_pairwise_affine_score={score}")
    print(f"python_pairwise_affine_matches={matches}")
    print(f"python_pairwise_affine_identity={identity:.6f}")
    print(f"python_pairwise_affine_length={len(aligned_left)}")


if __name__ == "__main__":
    main()
